### PR TITLE
store migrations: cache migrations and add note

### PR DESCRIPTION
**Pull request by 6543**
* Cache migrations instead of query each time
* Add a note of migration execute order

### CODE REVIEW

File: server/store/datastore/migration/002_repos_drop_allow_deploys_allow_tags.go [hunk 1]
```diff
@@ -26,4 +26,5 @@ var alterTableReposDropAllowDeploysAllowTags = task{
 			"repo_allow_tags",
 		)
 	},
+	dependency: []string{"xorm"},
```
Comment (anbraten) [inline server/store/datastore/migration/002_repos_drop_allow_deploys_allow_tags.go:29]: Wouldn't it be enough to say, migrations are executed in order an fail if one or more of the previous ones were skipped?

Comment (6543) [reply server/store/datastore/migration/002_repos_drop_allow_deploys_allow_tags.go:29]: currently if a migration fail, sql transaktion do a rollback and quit the program

**Comment by 6543** [comment]
I'll remove dependency logic and add a code comment instead ..

File: server/store/datastore/migration/migration.go [hunk 2]
```diff
@@ -25,6 +25,7 @@ import (
 )
 
 // APPEND NEW MIGRATIONS
+// they are executed in order and if one fail woodpecker try to rollback and quite
```
Comment (6543) [inline server/store/datastore/migration/migration.go:28]:
```suggestion
// they are executed in order and if one fail woodpecker try to rollback and quit
```

File: server/store/datastore/migration/migration.go [hunk 3]
```diff
@@ -92,13 +93,19 @@ func Migrate(e *xorm.Engine) error {
...
+
 	for _, task := range tasks {
 		log.Trace().Msgf("start migration task '%s'", task.name)
-		exist, err := sess.Exist(&migrations{task.name})
-		if err != nil {
-			return err
-		}
-		if exist {
+		if migCache[task.name] {
 			log.Trace().Msgf("migration task '%s' exist", task.name)
```
Comment (anbraten) [inline server/store/datastore/migration/migration.go:109]:
```suggestion
			log.Trace().Msgf("migration task '%s' already applied", task.name)
```

File: server/store/datastore/migration/migration.go [hunk 4]
```diff
@@ -92,13 +93,19 @@ func Migrate(e *xorm.Engine) error {
...
+	var migList []*migrations
+	if err := sess.Find(&migList); err != nil {
+		return err
+	}
+	for i := range migList {
+		migCache[migList[i].Name] = true
+	}
+
 	for _, task := range tasks {
 		log.Trace().Msgf("start migration task '%s'", task.name)
```
Comment (anbraten) [inline server/store/datastore/migration/migration.go:107]:
```suggestion
```

File: server/store/datastore/migration/migration.go [hunk 5]
```diff
@@ -92,13 +93,19 @@ func Migrate(e *xorm.Engine) error {
...
-		exist, err := sess.Exist(&migrations{task.name})
-		if err != nil {
-			return err
-		}
-		if exist {
+		if migCache[task.name] {
 			log.Trace().Msgf("migration task '%s' exist", task.name)
 			continue
 		}
```
Comment (anbraten) [inline server/store/datastore/migration/migration.go:111]:
```suggestion
		}

		log.Trace().Msgf("start migration task '%s'", task.name)
```

**Review by anbraten — APPROVED** [review]

**Pull request merged**